File: src/core/Options/Expire.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using core.Shared;
using MediatR;

namespace core.Options
{
    public class Expire
    {
        public class Command : RequestWithUserId<CommandResponse>
        {
            [Required]
            public Guid? Id { get; set; }
        }

        public class Handler : IRequestHandler<Command, CommandResponse>
        {
            private IPortfolioStorage _storage;

            public Handler(IPortfolioStorage storage)
            {
                _storage = storage;
            }
            
            public async Task<CommandResponse> Handle(Command cmd, CancellationToken cancellationToken)
            {
                var option = await _storage.GetOwnedOption(cmd.Id.Value, cmd.UserId);
                if (option == null)
                {
                    return CommandResponse.Failed("Trying to expire not owned option");
                }

                option.Expire();

                await this._storage.Save(option, cmd.UserId);

                return CommandResponse.Success();
            }
        }
    }
}

File: src/core.fs/Options/Expire.fs
module core.fs.Options.Expire

    open System
    open core.Options
    open core.Shared
    open core.Shared.Adapters.Storage
    open core.fs
    
    type ExpireData(optionId: Guid, userId: Guid) =
        member this.OptionId = optionId
        member this.UserId = userId
        
    type Command =
        | Expire of ExpireData
        | Assign of ExpireData
        
    type ExpireViaLookupData(ticker:string, strikePrice:decimal, expiration:DateTimeOffset, userId: Guid) =
        member this.Ticker = ticker
        member this.StrikePrice = strikePrice
        member this.Expiration = expiration
        member this.UserId = userId
        
    type LookupCommand =
        | ExpireViaLookup of ExpireViaLookupData
        | AssignViaLookup of ExpireViaLookupData
                
    type Handler(storage: IPortfolioStorage) =
        
        let expireOption command = task {
            let data,assign = 
                match command with
                | Expire(data) -> (data, false)
                | Assign(data) -> (data, true)
                
            let! option = storage.GetOwnedOption(optionId = data.OptionId, userId = data.UserId)
            
            match option with
            | null ->
                return $"option for id {data.OptionId} not found" |> ResponseUtils.failedTyped<OwnedOption>
            | _ ->
                option.Expire(assign=assign)
                do! storage.Save(option, data.UserId)
                return ServiceResponse<OwnedOption>(option)
        }
        
        let expireViaLookup (command:LookupCommand) = task {
            let data,assigned = 
                match command with
                | ExpireViaLookup(data) -> (data, false)
                | AssignViaLookup(data) -> (data, true)
                
            let! options = storage.GetOwnedOptions(data.UserId)
            let option =
                options
                |> Seq.tryFind(fun o -> o.State.Ticker = data.Ticker && o.State.StrikePrice = data.StrikePrice && o.State.Expiration = data.Expiration)
                
            match option with
            | Some(o) ->
                o.Expire(assign=assigned)
                do! storage.Save(o, data.UserId)
                return ServiceResponse<OwnedOption>(o)
                
            | None ->
                return $"option for ticker {data.Ticker} strike {data.StrikePrice} expiration {data.Expiration} not found" |> ResponseUtils.failedTyped<OwnedOption>
        }
        
        interface IApplicationService
        
        member this.Handle(command: Command) = command |> expireOption
            
        member this.Handle(command: LookupCommand) = command |> expireViaLookup
            